### PR TITLE
feature/1 added runtime configuration of CORS

### PR DESCRIPTION
**Pull request by captaincoordinates**
**Related Issue(s):** #
#1, https://github.com/sparkgeo/prescient/issues/140 

**Description:**
Support CORS configuration via JSON file and environment variable identifying file location.
Additional tests to support CORS testing.
Test infrastructure updates to support executing a setup function before fixtures, necessary as middleware configuration logic is executed as part of fixtures but new tests require pre-middleware setup.

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/master/CHANGES.md).

### CODE REVIEW

File: stac_fastapi/pgstac/tests/api/test_api.py [hunk 1]
```diff
@@ -302,3 +316,62 @@ async def test_search_line_string_intersects(
...
+        )
+        == 0
+    )
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("app_client", [{"setup_func": cors_missing}], indirect=True)
+async def test_with_missing_config(app_client):
+    resp = await app_client.get("/", headers={"Origin": cors_permit_origin})
+    assert resp.status_code == HTTPStatus.OK
```
Comment (aspcanada) [inline stac_fastapi/pgstac/tests/api/test_api.py:370]: If you're missing the config, you want this to pass with http 200?

File: stac_fastapi/pgstac/tests/api/test_api.py [hunk 2]
```diff
@@ -302,3 +316,62 @@ async def test_search_line_string_intersects(
...
+    assert resp.headers["access-control-allow-origin"] == cors_permit_origin
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("app_client", [{"setup_func": cors_enable}], indirect=True)
+async def test_with_mismatch_cors(app_client):
+    resp = await app_client.get("/", headers={"Origin": cors_deny_origin})
+    assert resp.status_code == HTTPStatus.OK
```
Comment (aspcanada) [inline stac_fastapi/pgstac/tests/api/test_api.py:353]: does a mismatched cors origin still respond with 200 OK?

File: stac_fastapi/sqlalchemy/tests/api/test_api.py [hunk 3]
```diff
@@ -285,3 +301,58 @@ def test_search_line_string_intersects(
...
+            ]
+        )
+        == 0
+    )
+
+
+@pytest.mark.parametrize("app_client", [{"setup_func": cors_missing}], indirect=True)
+def test_with_missing_config(app_client):
+    resp = app_client.get("/", headers={"Origin": cors_permit_origin})
+    assert resp.status_code == HTTPStatus.OK
```
Comment (aspcanada) [inline stac_fastapi/sqlalchemy/tests/api/test_api.py:351]: same as above...

File: stac_fastapi/sqlalchemy/tests/api/test_api.py [hunk 4]
```diff
@@ -285,3 +301,58 @@ def test_search_line_string_intersects(
...
+def test_with_match_cors(app_client):
+    resp = app_client.get("/", headers={"Origin": cors_permit_origin})
+    assert resp.status_code == HTTPStatus.OK
+    assert resp.headers["access-control-allow-origin"] == cors_permit_origin
+
+
+@pytest.mark.parametrize("app_client", [{"setup_func": cors_enable}], indirect=True)
+def test_with_mismatch_cors(app_client):
+    resp = app_client.get("/", headers={"Origin": cors_deny_origin})
+    assert resp.status_code == HTTPStatus.OK
```
Comment (aspcanada) [inline stac_fastapi/sqlalchemy/tests/api/test_api.py:335]: same as above...

File: stac_fastapi/pgstac/tests/api/test_api.py (hunk 1 above)
Comment (captaincoordinates) [reply stac_fastapi/pgstac/tests/api/test_api.py:370]: That's how CORS works - if a browser submits an HTTP request from a domain not listed in CORS the request is successful but does not include the right headers. The browser then prevents JavaScript from accessing data from the response. It's actually infuriating if you're a web developer as you can see the request succeed in your network traffic but then you don't get any data 🤷 . CORS is all about bypassing the browser's same-origin policy; it's not about server security or authorisation: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

Missing config is configured to behave the same as no config, for the least intrusive impact on the API

File: stac_fastapi/pgstac/tests/api/test_api.py (hunk 2 above)
Comment (captaincoordinates) [reply stac_fastapi/pgstac/tests/api/test_api.py:353]: Yes, see above

File: stac_fastapi/sqlalchemy/tests/api/test_api.py (hunk 3 above)
Comment (captaincoordinates) [reply stac_fastapi/sqlalchemy/tests/api/test_api.py:351]: Yes, see above

File: stac_fastapi/sqlalchemy/tests/api/test_api.py (hunk 4 above)
Comment (captaincoordinates) [reply stac_fastapi/sqlalchemy/tests/api/test_api.py:335]: Yes, see above

File: stac_fastapi/api/stac_fastapi/api/middleware.py [hunk 5]
```diff
@@ -29,3 +35,50 @@ async def _middleware(request: Request, call_next):
...
+    )
+    if not has_cors_middleware:
+        cors_config_location_key: Final = "CORS_CONFIG_LOCATION"
+        if cors_config_location_key in environ:
+            cors_config_path = environ[cors_config_location_key]
+            logger.info(f"looking for CORS config file at {cors_config_path}")
+            if path.exists(cors_config_path):
+                try:
+                    with open(cors_config_path, "r") as cors_config_file:
+                        cors_config = loads("".join(cors_config_file.readlines()))
```
Comment (ghelobytes) [inline stac_fastapi/api/stac_fastapi/api/middleware.py]: Can this be simplified to `cors_config = loads(cors_config_file.read())`?

File: stac_fastapi/api/stac_fastapi/api/middleware.py [hunk 6]
```diff
@@ -29,3 +35,50 @@ async def _middleware(request: Request, call_next):
...
+                        )
+                        logger.debug(f"loaded CORS config {cors_config}")
+                except ValueError as e:
+                    logger.error(f"error parsing JSON at {cors_config_path}: {e}")
+                except OSError as e:
+                    logger.error(f"error reading {cors_config_path}: {e}")
+            else:
+                logger.warning(f"CORS config not found at {cors_config_path}")
+
+    return extended_middlewares
```
Comment (cuttlefish) [inline stac_fastapi/api/stac_fastapi/api/middleware.py]: Do we want to warn the user if they already have CORS middleware loaded and also have a CORS config file?

File: stac_fastapi/api/stac_fastapi/api/middleware.py (hunk 5 above)
Comment (captaincoordinates) [reply stac_fastapi/api/stac_fastapi/api/middleware.py]: it sure can

File: stac_fastapi/api/stac_fastapi/api/middleware.py (hunk 6 above)
Comment (captaincoordinates) [reply stac_fastapi/api/stac_fastapi/api/middleware.py]: sure, working on it

Comment (captaincoordinates) [reply stac_fastapi/api/stac_fastapi/api/middleware.py]: good job you mentioned this, I added a new test for it and caught a bug in my implementation...

**Comment by captaincoordinates** [comment]
Dealing with @cuttlefish feedback I identified a bug that was a hangover from a previous implementation approach. Added a new test for this.
Also refactored the middleware module to better accommodate additional runtime configurations in future.

**Review by cuttlefish — APPROVED** [review]

**Review by ghelobytes — APPROVED** [review]

**Pull request merged**